### PR TITLE
Event-specific subject and page properties

### PR DESCRIPTION
**Pull request by mscwilson**
For issue #132.

All the `track_x` methods now allow an optional event-specific Subject (for setting userid, screen size etc), and/or Page. Page is a new class that stores URL, referrer URL, and page title.

These event-specific parameters will save data into the event's "atomic" keys/eventual table column headings, according to the [Snowplow tracker protocol](https://docs.snowplowanalytics.com/docs/collecting-data/collecting-from-own-applications/snowplow-tracker-protocol). This is in contrast to context entities, which are stored as JSON in the 'context' column.

Settings in the event Subject override tracker Subject settings. Similarly, for `track_page_view`, an e.g. page title defined as a property of the Page object will override a page title provided as argument.

### CODE REVIEW

**Comment by snowplowcla** [comment]
Thanks for your pull request. Is this your first contribution to a Snowplow open source project? Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

:memo: Please visit https://docs.snowplowanalytics.com/docs/contributing/contributor-license-agreement/ to learn more and sign.

Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.

File: lib/snowplow-tracker/tracker.rb [hunk 1]
```diff
@@ -122,38 +120,58 @@ def build_context(context)
 
     # Tracking methods
 
-    # Attaches all the fields in @standard_nv_pairs to the request
+    # Attaches all the fields in @settings to the request
     # Only attaches the context vendor if the event has a custom context
```
Comment (mscwilson) [inline lib/snowplow-tracker/tracker.rb:124]: is this true?

File: lib/snowplow-tracker/tracker.rb [hunk 2]
```diff
@@ -122,38 +120,58 @@ def build_context(context)
...
       nil
     end
 
+    Contract Or[Timestamp, Num, nil] => Timestamp
+    def process_tstamp(tstamp)
+      tstamp = Timestamp.create if tstamp.nil?
+      tstamp = DeviceTimestamp.new(tstamp) if tstamp.is_a? Numeric
+      tstamp
+    end
+
```
Comment (paulboocock) [inline lib/snowplow-tracker/tracker.rb:139]: I'm just looking at this and wondering if it's historically incorrect. With the introduction of `true_timestamp` we give users two timestamps to play with. Ideally users shouldn't really set the device timestamp (`dtm`) but we continue letting them. However, if the true timestamp is set, then the `dtm` property should still be set. If I'm reading this correctly, if the user passes a `ttm` (TrueTimestamp) into the `tstamp` property, then we won't set the `dtm` field. I think we could update `finalise_payload` to still set `dtm` (to current time) if `ttm` is set.

File: lib/snowplow-tracker/tracker.rb (hunk 1 above)
Comment (mscwilson) [reply lib/snowplow-tracker/tracker.rb:124]: I was confused by `emitter.input(payload.context)`. But Payload stores the event details in a hash called context, which is unrelated to the event context (Payload.context might contain the 'co' key, if event context had been given).

Is there any reason for this name or can I change it?

Comment (paulboocock) [reply lib/snowplow-tracker/tracker.rb:124]: Seems strange that the payload is in a property called `context` for sure. `payload.data` is a more popular name across the estate. `co` is the non-base64 equivalent to `cx` where the entity/context array ends up, I don't think there's much relation to this name so I think changing it is less confusing.

**Review by paulboocock — APPROVED** [review]

**Pull request merged**